Memtx tree pagination

drewdzzz opened the pull request:
This patchset adds helpers, needed for pagination, new options after and fetch_pos for select (both FFI and LuaC), and pagination to memtx tree. All the types of tree indexes (multikey, functional) are suported.

Documentation will be added in a later patchset.

Closes #7633
Closes #7635
Closes #7636

locker commented on src/box/box.cc:
```diff
@@ -2567,15 +2585,26 @@ box_select(uint32_t space_id, uint32_t index_id,
+	 * it must retain empty, and if no tuple greater then tuple with
+	 * position pos, we must retain it too.
+	 */
+	if (update_pos && index_iterator_position(index, it, &pos->key,
```
The index may be dropped in the meantime. We need to handle this. Please take a look at `iterator_is_valid`. Let's raise an error in this case.

drewdzzz replied:
done

locker commented on src/box/box.cc:
```diff
@@ -2519,6 +2531,12 @@ box_select(uint32_t space_id, uint32_t index_id,
 		return -1;
+	const char *after = pos != NULL ? pos->key : NULL;
+	if (after != NULL) {
+		uint32_t after_size = mp_decode_array(&after);
```
We need to check that `after` is an `MP_ARRAY`.

drewdzzz replied:
Stale since we decided to unpack position in `box_select_after`.

locker commented on src/box/lua/misc.cc:
```diff
@@ -61,6 +63,29 @@ lbox_encode_tuple_on_gc(lua_State *L, int idx, size_t *p_len)
+	struct errinj *inj = errinj(ERRINJ_POSITION_PACK_ON_GC_ALLOC,
+				    ERRINJ_BOOL);
+	if (inj == NULL || inj->bparam == false)
+		buf = (char *)region_alloc(&fiber()->gc, *size);
```
Looks like you don't clean up the region after usage. The fiber region isn't cleaned up automatically - it should be cleaned up explicitly with `region_truncate`.

drewdzzz replied:
done (in LuaC too)

locker commented on src/box/index.cc:
```diff
@@ -153,6 +154,16 @@ exact_key_validate(struct key_def *key_def, const char *key,
 }
 
+int
+position_validate(const struct index_def *index_def, const char *key,
```
Let's drop this method and call `exact_key_validate` directly.

drewdzzz replied:
done

locker commented on src/box/lua/misc.cc:
```diff
@@ -61,6 +63,29 @@ lbox_encode_tuple_on_gc(lua_State *L, int idx, size_t *p_len)
 }
 
+API_EXPORT const char *
+position_pack_on_gc(struct position *pos, uint32_t *size)
```
Since `box_select_after` returns a packed position, this function won't be needed anymore.

drewdzzz replied:
done

locker commented on src/box/memtx_tree.cc:
```diff
@@ -932,6 +965,138 @@ memtx_tree_index_get_internal(struct index *base, const char *key,
+ * Version of memtx_tree_index_iterator_position for multikey index.
+ */
+static int
+memtx_tree_index_iterator_position_multikey(struct index *base,
```
Let's join it with `memtx_tree_index_iterator_position` using template parameter.

drewdzzz replied:
done

locker commented on src/box/memtx_tree.cc:
```diff
@@ -932,6 +965,138 @@ memtx_tree_index_get_internal(struct index *base, const char *key,
+	assert(pos != NULL);
+	assert(size != NULL);
+
+	struct errinj *inj = errinj(ERRINJ_ITERATOR_POSITION_FAIL,
```
Let's drop this error injection.

drewdzzz replied:
done

locker commented on src/box/tuple_compare.cc:
```diff
@@ -1447,17 +1447,50 @@ func_index_compare_with_key(struct tuple *tuple, hint_t tuple_hint,
+	const char *field;
+	/* Skip compared fields. */
+	for (uint32_t i = 0; i < cmp_part_count; i++) {
+		mp_next(&key);
```
Double scan of the key looks sub-optimal. Please refactor so that we continue scan from where we stopped.

Also, please add a unit test, see [this](https://github.com/tarantool/tarantool-ee/blob/master/test/enterprise-unit/key_def_raw.c) for example.

drewdzzz replied:
done

locker commented on src/box/lua/schema.lua:
```diff
@@ -114,12 +114,37 @@ ffi.cdef[[
+                     int iterator, uint32_t offset, uint32_t limit,
+                     const char *key, const char *key_end,
+                     struct position *pos, bool update_pos, struct port *port);
+
```
Instead of adding `box_select_after`, let's add new arguments to `box_select`. Just in case, please ask @Totktonada if we can change it.

drewdzzz replied:
Done. We will ask him after his vacation.

drewdzzz commented on src/box/index.h:
```diff
@@ -479,6 +499,14 @@ struct index_vtab {
+	 * tuple. If there were no tuples fetched, pos and size are not
+	 * updated.
+	 */
+	int (*iterator_position)(struct index *index, struct iterator *it,
```
done

drewdzzz commented on src/box/box.h:
```diff
@@ -331,6 +332,18 @@ box_select(uint32_t space_id, uint32_t index_id,
+ * Pre-requesites: if update_pos is true, pos must not be NULL.
+ * box_select_after is private and used only by FFI.
+ */
+API_EXPORT int
```
deleted API_EXPORT

drewdzzz commented on src/box/schema.cc:
```diff
@@ -168,7 +168,8 @@ schema_find_id(uint32_t system_space_id, uint32_t index_id,
 	mp_encode_str(key, name, len);
-	struct iterator *it = index_create_iterator(index, ITER_EQ, key, 1);
+	struct iterator *it = index_create_iterator(index, ITER_EQ, key, 1,
+						    NULL);
```
done

drewdzzz commented on src/box/lua/schema.lua:
```diff
@@ -2443,20 +2486,49 @@ local function check_select_safety(index, key_is_nil, itype, limit, offset,
+    end
+end
+
 base_index_mt.select_ffi = function(index, key, opts)
```
done

drewdzzz commented on src/box/index.h:
```diff
@@ -494,8 +522,8 @@ struct index_vtab {
-			enum iterator_type type,
-			const char *key, uint32_t part_count);
+			enum iterator_type type, const char *key,
+			uint32_t part_count, const char *after);
```
done

drewdzzz commented on src/box/lua/schema.lua:
```diff
@@ -2443,20 +2486,49 @@ local function check_select_safety(index, key_is_nil, itype, limit, offset,
+    if after == nil then
+        builtin.position_reset(position)
+    elseif type(after) == "string" then
+        nok = builtin.position_unpack(after, position) ~= 0
```
done

drewdzzz commented on src/box/box.h:
```diff
@@ -331,6 +332,18 @@ box_select(uint32_t space_id, uint32_t index_id,
+box_select_after(uint32_t space_id, uint32_t index_id,
+		 int iterator, uint32_t offset, uint32_t limit,
+		 const char *key, const char *key_end,
+		 struct position *pos, bool update_pos, struct port *port);
```
done

drewdzzz commented on src/box/position.h:
```diff
@@ -0,0 +1,38 @@
+ * more or equal than packed size of position.
+ */
+void
+position_pack(struct position *pos, char *buffer);
```
done

drewdzzz commented on src/box/sysview.c:
```diff
@@ -110,10 +110,15 @@ sysview_index_destroy(struct index *index)
+	if (unlikely(after != NULL)) {
+		diag_set(UnsupportedIndexFeature, base->def, "pagination");
+		return NULL;
+	}
```
done

locker commented on src/box/lua/schema.lua:
```diff
@@ -2502,6 +2561,23 @@ base_index_mt.alter = function(index, options)
+    if nok then
+        box.error()
+    end
+    return ffi.string(pos[0], pos_end[0] - pos[0])
```
I think you should clean up the region here, as well.

Please add a test that runs `select` and `position` Lua functions in a loop and checks that `box.runtime.info().used` doesn't grow without bounds.

drewdzzz replied:
Done

locker commented on src/box/memtx_tree.cc:
```diff
@@ -709,6 +736,13 @@ tree_iterator_start(struct iterator *iterator, struct tuple **ret)
 		res = memtx_tree_iterator_get_elem(tree, &it->tree_iterator);
 	}
+	/* If we skip tuple, flag equals is not actual - need to refresh it. */
+	if (skip_eq && res != NULL && (type == ITER_EQ || type == ITER_REQ) &&
```
Confusing variable name - `skip_eq`. Sounds like if it's set, the EQ check should be skipped, but actually it's vice versa. Better invert its meaning?

drewdzzz replied:
Renamed to `skip_equal_tuple` and use it to change `GE` to `GT` and `LE` to `LT`

locker commented on src/box/memtx_tree.cc:
```diff
@@ -930,6 +965,120 @@ memtx_tree_index_get_internal(struct index *base, const char *key,
+/**
+ * Implementation of @tree_iterator_position for general and multikey indexes.
+ */
+template <bool USE_HINT, bool is_multikey>
```
Nit: All template variables in this file should be in CAPS for consistency.

locker commented on src/box/memtx_tree.cc:
```diff
@@ -930,6 +965,120 @@ memtx_tree_index_get_internal(struct index *base, const char *key,
 }
 
+/**
+ * Implementation of @tree_iterator_position for general and multikey indexes.
```
Nit: Please don't use `@` or other Doxygen-like notation in internal comments. They are useless and only make the comments look inconsistent, because we don't usually use them.

drewdzzz replied:
Removed all doxygen-style comments from the patch.

locker commented on src/box/tuple_compare.cc:
```diff
@@ -1447,17 +1469,48 @@ func_index_compare_with_key(struct tuple *tuple, hint_t tuple_hint,
-					      key_def);
+	for (uint32_t i = cmp_part_count; i < part_count; i++) {
+		struct key_part *part = &key_def->parts[i];
+		field = tuple_field_raw_by_part(format, tuple_raw, field_map,
```
Please also prepare a patch for [`key_def_raw.c`](https://github.com/tarantool/tarantool-ee/blob/a74aeaf7c6f1526a3aa723ed443f620eeb2cc265/src/box/key_def_raw.c#L242-L247) from the EE repository: currently, it assumes that the function index implementation of `tuple_compare_with_key` doesn't access tuple format, but it isn't true anymore so we need to provide an override.

locker replied:
The function may actually be used by a read view in Lua so a stub isn't an option. An override that uses functional key parts only and has an assertion that an extended key definition (with primary key parts) is never used should be fine. I opened a ticket and assigned it to you: https://github.com/tarantool/tarantool-ee/issues/248.

drewdzzz replied:
[Done](https://github.com/tarantool/tarantool-ee/pull/250).

locker commented on src/box/memtx_tree.cc:
```diff
@@ -930,6 +965,120 @@ memtx_tree_index_get_internal(struct index *base, const char *key,
+	*size = alloc_size;
+	*pos = data;
+	/* Pack an array with concatenated func key and primary key. */
+	memcpy(data, func_key, func_key_size);
```
I think you should implement `tuple_extract_key` for functional indexes. Could be done in a separate patch with a unit test, like you did with `tuple_compare_with_key`.

drewdzzz replied:
We need to pass func key to `tuple_extract_key`, but there is no argument in signature of the faimly of extracting funcitons. We have two options:
1. Extend signature of method (or use int64_t instead of int for multikey_idx, then we could pass pointer as int_64) or add new method `extract_key_func` to `key_def`.
2. Implement separate method for functional index, but it won't be virtual method of `key_def`.

locker replied:
I see. I'm inclined to think that we should replace `int multikey_idx` argument with `hint_t`, but I'm not sure. We need to discuss this with @alyapunov first.

locker replied:
Discussed with @alyapunov. Decided that it's best to leave it as is, because in future we'll probably  store functional index key parts right in tuples (in hidden fields), which would make passing `hint_t` to `extract_key` pointless. Nothing to do here.

locker commented on src/box/memtx_tree.cc:
```diff
@@ -930,6 +965,120 @@ memtx_tree_index_get_internal(struct index *base, const char *key,
+
+	struct key_def *key_def = it->index->def->key_def;
+	int rc = -1;
+	if (key_def->for_func_index) {
```
`iterator_position` is a virtual method so we should set it to the right function pointer instead of doing branching inside.

drewdzzz replied:
Moved this branching to `memtx_tree_create_iterator` method (since it is virtual method of iterator, we cannot get rid of branching at all).

locker commented on src/box/memtx_tree.cc:
```diff
@@ -1431,9 +1577,6 @@ memtx_tree_index_create_iterator_after(struct index *base,
 
-	if (type == ITER_ALL)
-		type = ITER_GE;
-
```
Looks like unnecessary change.

drewdzzz replied:
Removed the change.

locker commented on src/box/memtx_tree.cc:
```diff
@@ -1457,6 +1600,26 @@ memtx_tree_index_create_iterator_after(struct index *base,
+			 */
+			it->type = iterator_type_is_reverse(it->type) ?
+						 ITER_LT :
+						 ITER_GT;
```
I don't like that we overwrite `it->type` for some iterator types (le, ge), but not for others (eq, req). Can you avoid that?

drewdzzz replied:
We can overwrite `GE` and 'LE` inside `tree_iterator_start` or move this logic to https://github.com/tarantool/tarantool/blob/dc147ec9190fc8bbbe56ce8a4a1718fbf8b03484/src/box/memtx_tree.cc#L681-L689

locker replied:
Right, I think we'd better do that in `tree_iterator_start` - this way `tree_iterator->key` and `tree_iterator->type` would denote the original iteration key and direction.

drewdzzz replied:
Moved changing range iterators with equality to `tree_iterator_start`, `tree_iterator->key` isn't affected anymore.

drewdzzz commented on src/box/index.h:
```diff
@@ -493,10 +511,16 @@ struct index_vtab {
+	 * position, passed in after argument. Argument after is an extracted
+	 * cmp_def without MP_ARRAY header.
+	 */
+	struct iterator *(*create_iterator_after)(struct index *index,
```
It may be a bit confusing, but okay, let's do that

drewdzzz replied:
done

drewdzzz commented on src/box/memtx_bitset.cc:
```diff
@@ -334,15 +334,22 @@ memtx_bitset_index_replace(struct index *base, struct tuple *old_tuple,
 	return 0;
 }
 
+/** Implementation of create_iteartor_after for memtx bitset index. */
```
fixed

drewdzzz commented on src/box/memtx_bitset.cc:
```diff
@@ -334,15 +334,22 @@ memtx_bitset_index_replace(struct index *base, struct tuple *old_tuple,
 
 	assert(part_count == 0 || key != NULL);
 	(void) part_count;
+	if (unlikely(after != NULL)) {
```
done

drewdzzz commented on src/box/errcode.h:
```diff
@@ -298,6 +298,7 @@ struct errcode_record {
 	/*243 */_(ER_SSL,			"%s") \
 	/*244 */_(ER_SPLIT_BRAIN,		"Split-Brain discovered: %s") \
 	/*245 */_(ER_OLD_TERM,			"The term is outdated: old - %llu, new - %llu") \
+	/*246 */_(ER_INVALID_POSITION,		"Position descriptor is invalid") \
```
done

drewdzzz commented on src/box/index.cc:
```diff
@@ -573,6 +573,96 @@ iterator_position(struct iterator *it, const char **pos, uint32_t *size)
 
+/** Keys for position map. All the keys must be uint. */
+enum iterator_position_keys {
+	POSITION_KEY,
```
done

drewdzzz commented on src/box/index.cc:
```diff
@@ -573,6 +573,96 @@ iterator_position(struct iterator *it, const char **pos, uint32_t *size)
+iterator_position_pack_size(const char *pos, const char *pos_end,
+			    uint32_t part_count)
+{
+	if (pos == NULL)
```
Yep, assert will be better.
Replaced `if` with `assert`.

drewdzzz commented on src/box/index.cc:
```diff
@@ -573,6 +573,96 @@ iterator_position(struct iterator *it, const char **pos, uint32_t *size)
+			array = data;
+			*part_count = mp_decode_array(&data);
+			*pos = data;
+			if (mp_check(&array, data_end) != 0)
```
Added the check and corresponding test suite.

drewdzzz commented on test/unit/iterator_position.c:
```diff
@@ -0,0 +1,151 @@
+
+char key_buf[KEY_BUF_LEN];
+const char *key_buf_no_header;
+char pos_buf[POS_BUF_LEN];
```
Added static qualifier.

drewdzzz commented on test/unit/key_def.c:
```diff
@@ -0,0 +1,258 @@
+		test_tuple_new("[%u%s%u%u%s]", 20, "foo", 100, 10, "cpp");
+	tuple->format_id = tuple_format_runtime->id;
+	func_key->format_id = tuple_format_runtime->id;
+	model->format_id = tuple_format_runtime->id;
```
fixed

drewdzzz commented on test/unit/key_def.c:
```diff
@@ -0,0 +1,258 @@
+	for (unsigned int i = 0; i < lengthof(keys); i++)
+		free(keys[i]);
+	tuple_delete(func_key);
+	key_def_delete(def);
```
fixed

drewdzzz commented on src/box/memtx_tree.cc:
```diff
@@ -437,13 +437,13 @@ tree_iterator_prev_base(struct iterator *iterator, struct tuple **ret)
 		memtx_tree_iterator_get_elem(&index->tree, &it->tree_iterator);
-	tree_iterator_set_current<USE_HINT>(it, res);
-	*ret = it->current.tuple;
+	*ret = res != NULL ? res->tuple : NULL;
```
`current` is set only in the case of exact match - if we have exact match on lower bound, it is not `NULL`.
https://github.com/tarantool/tarantool/blob/dc147ec9190fc8bbbe56ce8a4a1718fbf8b03484/src/box/memtx_tree.cc#L357-L378

drewdzzz replied:
I'll leave a comment there.

locker replied:
Looks like you're right. Let's add an assertion to `tree_iterator_set_current` that checks that neither `cur` nor `cur->tuple` can be `NULL`.

drewdzzz replied:
Good idea, I did so.

drewdzzz commented on src/box/memtx_tree.cc:
```diff
@@ -395,13 +395,13 @@ tree_iterator_next_base(struct iterator *iterator, struct tuple **ret)
 	if (*ret == NULL) {
 		iterator->next_internal = exhausted_iterator_next;
 	} else {
+		tree_iterator_set_current<USE_HINT>(it, res);
```
I forgot, that we decided to rename `current` to `last`. I renamed the field and added a comment.

drewdzzz commented on src/box/box.cc:
```diff
@@ -2481,6 +2483,19 @@ box_select(uint32_t space_id, uint32_t index_id,
 	if (key_validate(index->def, type, key, part_count))
 		return -1;
+	const char *pos, *pos_end;
+	uint32_t pc;
```
done

drewdzzz commented on src/box/lua/misc.cc:
```diff
@@ -61,6 +61,26 @@ lbox_encode_tuple_on_gc(lua_State *L, int idx, size_t *p_len)
+		if (box_index_tuple_position(space_id, index_id, tuple,
+					     tuple + size, pos, pos_end) != 0)
+			luaT_error(L);
+	} else { /* box.NULL */
```
Reworked the snippet, an error is raised if `after` has invalid type.

drewdzzz commented on src/box/lua/misc.cc:
```diff
@@ -183,12 +203,14 @@ port_msgpack_dump_lua(struct port *base, struct lua_State *L, bool is_flat)
-	if (lua_gettop(L) != 6 || !lua_isnumber(L, 1) || !lua_isnumber(L, 2) ||
-		!lua_isnumber(L, 3) || !lua_isnumber(L, 4) || !lua_isnumber(L, 5)) {
+	if (lua_gettop(L) != 8 || !lua_isnumber(L, 1) || !lua_isnumber(L, 2) ||
+	    !lua_isnumber(L, 3) || !lua_isnumber(L, 4) || !lua_isnumber(L, 5)) {
```
Added missing check.

drewdzzz commented on src/box/lua/schema.lua:
```diff
@@ -2430,20 +2454,52 @@ local function check_select_safety(index, key_is_nil, itype, limit, offset,
 end
 
+--[[ After must be a string, table or tuple, result is returned to a global
+     position instance. ]]--
```
Reworked the comment

drewdzzz commented on src/box/lua/misc.cc:
```diff
@@ -61,6 +61,26 @@ lbox_encode_tuple_on_gc(lua_State *L, int idx, size_t *p_len)
 
+void
+lbox_normalize_after(lua_State *L, int idx, int space_id, int index_id,
+		     const char **pos, const char **pos_end)
```
I named the function the same way as in `schema.lua`.

drewdzzz replied:
Since we decided to get rid of `after` in C, this method is called `lbox_normalize_position` now.

drewdzzz commented on src/box/index.h:
```diff
@@ -252,6 +252,14 @@ box_tuple_extract_key(box_tuple_t *tuple, uint32_t space_id,
 /** \endcond public */
 
+/**
+ * Get position of tuple in index (multikey and func indexes are not supported).
```
Only arguments that supposed to start iteration after position are called `after` (excluding some helpers like `normalize_after`). I don't find it confusing, but maybe I'm just used to it. Do you want to rename all the arguments in C, including argument of `create_iterator`?

locker replied:
Maybe, use `pos` for the argument passed and returned by iterator methods (`create_iterator`, `iterator_position`) and use `packed_pos` for the argument passed to `box_select` (i.e. for a packed position object)? Then there would be no confusion. Anyway, better discuss this with @alyapunov.

locker replied:
Discussed with @alyapunov. Agreed that we want to use the same names for variables denoting the same kind of thing **everywhere** to avoid confusion:
- Position key without headers: `pos, pos_end`.
- Packed position key: `packed_pos, packed_pos_end`.

`after` shouldn't be used at all.

drewdzzz replied:
Thank you! I removed all `after` arguments and variables in C, and renamed both helpers `normalize_after` to `normalize_position`.

locker commented on src/box/index.cc:
```diff
@@ -559,6 +603,113 @@ iterator_next_internal(struct iterator *it, struct tuple **ret)
+
+void
+iterator_position_pack(const char *pos, const char *pos_end,
+		       uint32_t part_count, char *buffer)
```
Better rename `buffer` to `packed_pos` and add `packed_pos_end` for consistency.

drewdzzz replied:
Done

drewdzzz commented on src/box/index.cc:
```diff
@@ -559,6 +603,113 @@ iterator_next_internal(struct iterator *it, struct tuple **ret)
+int
+iterator_position_unpack(const char **pos, const char **pos_end,
+			 uint32_t *part_count, const char *data,
+			 const char *data_end)
```
Done

locker commented on src/box/index.h:
```diff
@@ -252,6 +252,15 @@ box_tuple_extract_key(box_tuple_t *tuple, uint32_t space_id,
 /** \endcond public */
 
+/**
+ * Get packed position of tuple in index to pass it to box_select
```
Please mention that the position is allocated on the fiber region.

drewdzzz replied:
Done

locker commented on src/box/index.h:
```diff
@@ -252,6 +252,15 @@ box_tuple_extract_key(box_tuple_t *tuple, uint32_t space_id,
+int
+box_index_tuple_position(uint32_t space_id, uint32_t index_id,
+			 const char *tuple, const char *tuple_end,
+			 const char **pos, const char **pos_end);
```
We should use `packed_pos` here as well as in `lbox_select`, `lbox_normalize_position`, I think.
`pos` for packed position is okay to use only in Lua where there's no such thing as unpacked position.

drewdzzz replied:
Done

locker commented on src/box/lua/misc.cc:
```diff
@@ -61,6 +61,30 @@ lbox_encode_tuple_on_gc(lua_State *L, int idx, size_t *p_len)
 	return (char *) region_join_xc(gc, *p_len);
 }
 
+/** Convert packed position, tuple or table from Lua to C packed position. */
```
Again, mention that the position may be allocated on the fiber region so that whoever uses this code keeps in mind that the region should be cleaned up afterwards. Please mention it **everywehre** a position is returned.

locker commented on src/box/lua/misc.cc:
```diff
@@ -215,7 +243,13 @@ lbox_select(lua_State *L)
+		lua_pushlstring(L, pos, pos_end - pos);
+		ret_count++;
+	}
+	region_truncate(&fiber()->gc, svp);
```
You should clean up the region even if `box_select` above fails.

drewdzzz replied:
Done, also changed `lbox_normalize_position` - now it returns int instead of void to clean up the region if it fails.

drewdzzz commented on src/box/index.cc:
```diff
@@ -573,6 +573,108 @@ iterator_position(struct iterator *it, const char **pos, uint32_t *size)
+		case ITERATOR_POSITION_KEY:
+			if (packed_pos + 1 >= packed_pos_end ||
+			    mp_typeof(*packed_pos) != MP_ARRAY ||
+			    mp_check_array(packed_pos, packed_pos_end) > 0)
```
Anyway, we will need to check types (`MP_ARRAY`, etc.), plus `mp_check_*` functions looks really fast since they use only bitwise operators and `if`, while using `mp_check` we will unpack the whole MsgPack twice.

locker replied:
Discussed f2f. Decided to use `mp_check` results in cleaner code.

drewdzzz commented on src/box/memtx_tree.cc:
```diff
@@ -346,12 +356,13 @@ static void
 	struct tree_iterator<USE_HINT> *it = get_tree_iterator<USE_HINT>(iterator);
-	tree_iterator_set_current<USE_HINT>(it, NULL);
+	tree_iterator_set_last_tuple(it, NULL);
+	tree_iterator_set_last_hint(it, HINT_NONE);
```
Done.

drewdzzz commented on src/box/box.h:
```diff
@@ -324,12 +324,19 @@ box_demote(void);
+ * If packed_pos is not NULL and *packed_pos is not NULL, selection begins
+ * right after tuple with position, described by this argument.
+ * Pre-requesites: if update_pos is true, packed_pos and packed_pos_end must
+ * not be NULL.
```
Done

drewdzzz commented:
We found out that we need a test on interaction between `index:tuple_pos()` and option `after` of `select`.

drewdzzz commented:
Fixed a buf, added new test `test_tuple_pos_corner_cases` and added unique index to test `test_tree_pagination`.

drewdzzz commented:
Some problems with tuple position in nullable index.

locker commented on src/box/index.cc:
```diff
@@ -573,6 +573,109 @@ iterator_position(struct iterator *it, const char **pos, uint32_t *size)
+	if (packed_pos + bin_len != packed_pos_end)
+		goto fail;
+	packed_pos_check = packed_pos;
+	if (mp_check(&packed_pos_check, packed_pos_end) != 0)
```
Nit: Let's call `mp_check` before `mp_decode_binl`. Then you wouldn't need to use `mp_check_binl`.

drewdzzz replied:
Data that binary contains will not be checked then.

locker replied:
You'd still need the `mp_typeof(*packed_pos) != MP_BIN` check, but you wouldn't need the `packed_pos +1 >= packed_pos_end` and `mp_check_binl(packed_pos, packed_pos_end) != 0` checks. The code would be simpler and more consistent with the code below, where you don't use `mp_check_array`.

drewdzzz replied:
Done

locker replied:
Now, there's two `mp_check`... Also, you don't need to check that `packed_pos + bin_len == packed_pos_end` - instead you can just check that `packed_pos_check == packed_pos_end` after calling `mp_check`, which would be more straightforward.

locker replied:
Right.

locker commented on src/box/index.h:
```diff
@@ -364,6 +364,45 @@ iterator_next(struct iterator *it, struct tuple **ret);
+int
+iterator_position_unpack(const char *packed_pos, const char *packed_pos_end,
+			 const char **pos, const char **pos_end,
+			 uint32_t *part_count);
```
Come to think of it, the API looks asymmetrical: `pack` requires the packed position to be allocated on the region while `unpack` allocates a buffer by itself. I think we should either:
- Make `iterator_position_pack` allocate the position on the fiber region as well. I'm inclined to this one.
- Introduce `iterator_position_unpack_size` and require the caller to allocate the buffer.

drewdzzz replied:
`unpack` does not allocate memory - it points to the place where `iterator_position` is placed in passed packed position.

locker replied:
I missed that somehow, sorry about that. Then it's fine.

locker commented on test/unit/iterator_position.c:
```diff
@@ -0,0 +1,190 @@
+	ok(rc != 0, "Invalid position must not be unpacked");
+	mp_encode_strl(pos_buf + offset, key_len);
+	rc = iterator_position_unpack(pos_buf, pos_buf + pack_size,
+				      &begin, &end, &pos_part_count);
```
Nit: each test case should clean up the fiber region (which it uses for a position buffer) after a run.

locker replied:
Never mind - I thought that `unpack` allocates the result on the region while it, in fact, points to the location within the packed position, as you pointed out in the other comment.

locker commented on src/box/lua/schema.lua:
```diff
@@ -2502,6 +2561,26 @@ base_index_mt.alter = function(index, options)
+base_index_mt.tuple_pos = function(index, tuple)
+    check_index_arg(index, 'tuple_pos')
+    if not tuple then
+        box.error({reason="Usage index:tuple_pos(tuple)"})
```
Nit: We use `ER_PROC_LUA` in all other places in this source file to report about usage errors.

drewdzzz replied:
Done

locker commented on src/box/memtx_tree.cc:
```diff
@@ -356,7 +357,8 @@ static void
 	struct tree_iterator<USE_HINT> *it = get_tree_iterator<USE_HINT>(iterator);
-	tree_iterator_set_last_tuple(it, NULL);
+	if (it->last.tuple != NULL)
+		tuple_unref(it->last.tuple);
```
This and the changes in `tree_iterator_set_last_tuple` belong to the previous patch. Please rebase.

drewdzzz replied:
Done

drewdzzz commented:
- Added new commit that fixes a bug in `tuple_extract_key_slowpath_raw`.
- Added new helper `exact_key_validate_nullable` and use it in `box_select` instead of `exact_key_validate`.
- Added tests for nullable / exclude_null indexes.

locker commented on src/box/tuple_extract_key.cc:
```diff
@@ -254,21 +254,25 @@ tuple_extract_key_slowpath_raw(const char *data, const char *data_end,
+	uint32_t potential_null_count = key_def->part_count > field_count ?
+		key_def->part_count - field_count : 0;
+	uint32_t alloc_size = (uint32_t)(data_end - data) +
+		potential_null_count;
```
Let's simply add `key_def->part_count` bytes to the amount of memory allocated if `key_def->is_nullable` is set, irrespective of the field count. It'd be simpler, easier to understand, and more reliable. The key is allocated from temporary memory so allocating a few more bytes than necessary should be fine.

Also, please use `mp_sizeof_nil()` instead of implicitly assuming that the size of null is 1 byte.

drewdzzz replied:
Done

locker commented on test/unit/key_def.c:
```diff
@@ -223,13 +223,76 @@ test_func_compare_with_key(void)
+	key_def_delete(def_nullable_begin);
+	tuple_delete(tuple);
+	free(key_null_end);
+	free(key_null_begin);
```
Please add `fiber_gc` here - a test case should live no traces.

drewdzzz replied:
Removed explicit region_used/truncate and added fiber_gc - it is shorter and more concise.

locker commented on extra/exports:
```diff
@@ -86,6 +87,8 @@ box_latch_new
 box_latch_unlock
 box_on_shutdown
+box_read_ffi_disable
+box_read_ffi_enable
```
Nit: `box_read_ffi_enable` is never used from Lua so it shouldn't be exported.

drewdzzz replied:
Done

locker commented on test/box-luatest/pagination_test.lua:
```diff
@@ -0,0 +1,795 @@
+        local t = require('luatest')
+        local s = box.space.s
+        s:create_index("pk", {type = "tree"})
+        s:create_index("sk", {parts = {{field = 2, type = 'uint', path = '[*]'}}, type = "tree", unique=false})
```
Nit: The line's too long. Please check the file for long lines manually and fix them **all**.

drewdzzz replied:
Done

drewdzzz commented on src/box/index.cc:
```diff
@@ -573,6 +573,110 @@ iterator_position(struct iterator *it, const char **pos, uint32_t *size)
+			*pos_end = packed_pos;
+			break;
+		default:
+			goto fail;
```
Done

locker commented on src/box/tuple_extract_key.cc:
```diff
@@ -254,21 +254,25 @@ tuple_extract_key_slowpath_raw(const char *data, const char *data_end,
+			 "tuple_extract_key_raw");
+		return NULL;
+	}
+	char *key_buf = mp_encode_array(key, key_def->part_count);
```
Nit: Please always try to keep diff as possible, provided it doesn't hurt code readability. You don't need to move this hunk of code.

drewdzzz replied:
Fixed

locker commented on src/box/memtx_tree.cc:
```diff
@@ -346,12 +356,14 @@ static void
-	tree_iterator_set_current<USE_HINT>(it, NULL);
+	if (it->last.tuple != NULL)
+		tuple_unref(it->last.tuple);
+	tree_iterator_set_last_hint(it, HINT_NONE);
```
Please don't use `tree_iterator_set_last_hint` here - simply unref `last_func_key`, like you unref the last tuple.

drewdzzz replied:
Done

locker commented on src/box/memtx_tree.cc:
```diff
@@ -951,6 +991,91 @@ memtx_tree_index_get_internal(struct index *base, const char *key,
+ */
+template <bool USE_HINT, bool IS_MULTIKEY>
+static inline int
+tree_iterator_position(struct iterator *it, const char **pos, uint32_t *size)
```
Nit: this function is called by pointer so `inline` doesn't make any sense. Please remove.

locker commented on src/box/memtx_tree.cc:
```diff
@@ -951,6 +991,91 @@ memtx_tree_index_get_internal(struct index *base, const char *key,
+ * Implementation of iterator position for functional indexes.
+ */
+static inline int
+tree_iterator_position_func(struct iterator *it, const char **pos,
```
Ditto.

drewdzzz replied:
Done

locker commented on test/box-luatest/pagination_test.lua:
```diff
@@ -0,0 +1,801 @@
+local server = require('test.luatest_helpers.server')
+local t = require('luatest')
+local tree_g = t.group('Tree index tests',
+    t.helpers.matrix{engine={'memtx'}, disable_ffi={true, false}})
```
Please move this test to `engine-luatest`. We keep all multi-engine tests there for some reason. (Maybe, we shouldn't?)

drewdzzz replied:
I moved the test to `engine-luatest`

alyapunov reviewed (approved):
LGTM, great! the only thing that bothers me is `ffi.new('const char *[1]')` in every select.

drewdzzz commented on src/box/lua/schema.lua:
```diff
@@ -2430,20 +2454,52 @@ local function check_select_safety(index, key_is_nil, itype, limit, offset,
+-- begin and end of position.
+local function normalize_position(index, pos)
+    local nok
+    local ret_pos = ffi.new('const char *[1]')
```
We can make global instances of `pos` and `pos_end`

locker replied:
Done (amended the last commit).

<details><summary>diff</summary>

```diff
diff --git a/src/box/lua/schema.lua b/src/box/lua/schema.lua
index 66b8d4958c2c..92798426e8cf 100644
--- a/src/box/lua/schema.lua
+++ b/src/box/lua/schema.lua
@@ -2454,12 +2454,13 @@ local function check_select_safety(index, key_is_nil, itype, limit, offset,
     end
 end
 
+-- pointer to iterator position used by select() and tuple_pos()
+local iterator_pos = ffi.new('const char *[1]')
+local iterator_pos_end = ffi.new('const char *[1]')
+
 -- Argument pos must be a string, table or tuple, returns two C pointers:
 -- begin and end of position.
-local function normalize_position(index, pos)
-    local nok
-    local ret_pos = ffi.new('const char *[1]')
-    local ret_pos_end = ffi.new('const char *[1]')
+local function normalize_position(index, pos, ret_pos, ret_pos_end)
     if pos == nil then
         ret_pos[0] = nil
         ret_pos_end[0] = nil
@@ -2469,8 +2470,9 @@ local function normalize_position(index, pos)
     else
         local tuple_ibuf = cord_ibuf_take()
         local tuple, tuple_end = tuple_encode(tuple_ibuf, pos)
-        nok = builtin.box_index_tuple_position(index.space_id, index.id, tuple,
-                tuple_end, ret_pos, ret_pos_end) ~= 0
+        local nok = builtin.box_index_tuple_position(index.space_id, index.id,
+                                                     tuple, tuple_end,
+                                                     ret_pos, ret_pos_end) ~= 0
         cord_ibuf_put(tuple_ibuf)
         if nok then
             box.error()
@@ -2493,11 +2495,13 @@ base_index_mt.select_ffi = function(index, key, opts)
         check_select_opts(opts, key_is_nil)
     check_select_safety(index, key_is_nil, iterator, limit, offset, fullscan)
     local region_svp = builtin.box_region_used()
-    local pos, pos_end = normalize_position(index, after)
+    normalize_position(index, after, iterator_pos, iterator_pos_end)
     nok = builtin.box_select(index.space_id, index.id, iterator, offset, limit,
-            key, key_end, pos, pos_end, fetch_pos, port) ~= 0
-    if not nok and fetch_pos and pos[0] ~= nil then
-        new_position = ffi.string(pos[0], pos_end[0] - pos[0])
+                             key, key_end, iterator_pos, iterator_pos_end,
+                             fetch_pos, port) ~= 0
+    if not nok and fetch_pos and iterator_pos[0] ~= nil then
+        new_position = ffi.string(iterator_pos[0],
+                                  iterator_pos_end[0] - iterator_pos[0])
     end
     builtin.box_region_truncate(region_svp)
     cord_ibuf_put(ibuf)
@@ -2567,17 +2571,17 @@ base_index_mt.tuple_pos = function(index, tuple)
         box.error(box.error.PROC_LUA, "Usage index:tuple_pos(tuple)")
     end
     local region_svp = builtin.box_region_used()
-    local pos = ffi.new('const char *[1]')
-    local pos_end = ffi.new('const char *[1]')
     local ibuf = cord_ibuf_take()
     local data, data_end = tuple_encode(ibuf, tuple)
-    local nok = builtin.box_index_tuple_position(index.space_id, index.id, data,
-                                                data_end, pos, pos_end) ~= 0
+    local nok = builtin.box_index_tuple_position(index.space_id, index.id,
+                                                 data, data_end, iterator_pos,
+                                                 iterator_pos_end) ~= 0
     cord_ibuf_put(ibuf)
     if nok then
         box.error()
     end
-    local ret = ffi.string(pos[0], pos_end[0] - pos[0])
+    local ret = ffi.string(iterator_pos[0],
+                           iterator_pos_end[0] - iterator_pos[0])
     builtin.box_region_truncate(region_svp)
     return ret
 end
```

</details>

drewdzzz commented:
CI has found two problems:
1. Build is failed on gcc debug
2. `unit/key_def` fails on fedora

drewdzzz commented:
Build is fixed:
```diff
diff --git a/src/box/memtx_tree.cc b/src/box/memtx_tree.cc
index 567954e86..cbc5d94c1 100644
--- a/src/box/memtx_tree.cc
+++ b/src/box/memtx_tree.cc
@@ -1028,7 +1028,7 @@ tree_iterator_position(struct iterator *it, const char **pos, uint32_t *size)
 	}
 	int mk_idx = MULTIKEY_NONE;
 	if (IS_MULTIKEY)
-		mk_idx = tree_it->last.hint;
+		mk_idx = (int)tree_it->last.hint;
 	const char *key = tuple_extract_key(tuple, cmp_def, mk_idx, size);
 	if (key == NULL)
 		return -1;
 ```

locker commented:
I disabled Lua JIT in the test for aarch64 and filed a ticket https://github.com/tarantool/tarantool/issues/7739 to enable it back.
